nnunetv2/training/dataloading/pytorch_nnunet_dataset.py

```diff
@@ -0,0 +1,240 @@
+import os
+import shutil
+from typing import List, Tuple, Union
+
+import numpy as np
+
+from torch.utils.data import Dataset
+
+from nnunetv2.training.dataloading.utils import get_case_identifiers
+from nnunetv2.utilities.label_handling.label_handling import LabelManager
+
+import os
+
+
+# This has to be moved to an appropriate utils functions folder
+import pickle
+def load_pickle(file: str, mode: str = 'rb'):
+    with open(file, mode) as f:
+        a = pickle.load(f)
+    return a
+
+class nnUNetPytorchDataset(Dataset):
+    '''
+    Pytorch version of the nnUNet Dataset.
+    Goal is to implement the same functionality, but be able to use Pytorch's in-built
+    parallel dataloading abilities without having to use the Multi-Thread Augmenter from
+    batchgenerators.
+    For now, the only dependency we want on Batch Generators is 
+    (1) Transforms 
+
+    Next step -> I would want to also do away with dependency on Transforms
+
+    '''     
+    def __init__(self, folder: str, patch_size: Union[List[int], Tuple[int, ...], np.ndarray],
+                 final_patch_size: Union[List[int], Tuple[int, ...], np.ndarray],
+                 label_manager: LabelManager,
+                 transform,                 
+                 case_identifiers: List[str] = None,                 
+                 oversample_foreground_percent: float = 0.0,
+                 num_images_properties_loading_threshold: int = 2000,
+                 folder_with_segs_from_previous_stage: str = None):
+        # Initialize 
+        super().__init__()
+
+        self.final_patch_size = final_patch_size
+        self.patch_size = patch_size
+        self.need_to_pad = (np.array(patch_size) - np.array(final_patch_size)).astype(int)
+        self.has_ignore = label_manager.has_ignore_label
+        self.annotated_classes_key = tuple(label_manager.all_labels)
+        self.oversample_foreground_percent = oversample_foreground_percent
+        self.transform = transform
+
+        if case_identifiers is None:
+            case_identifiers = get_case_identifiers(folder)
+        case_identifiers.sort()
+
+        self.dataset = {}
+        for c in case_identifiers:
+            self.dataset[c] = {}
+            self.dataset[c]['data_file'] = os.path.join(folder, f"{c}.npz")
+            self.dataset[c]['properties_file'] = os.path.join(folder, f"{c}.pkl")
+            if folder_with_segs_from_previous_stage is not None:
+                self.dataset[c]['seg_from_prev_stage_file'] = os.path.join(folder_with_segs_from_previous_stage, f"{c}.npz")
+
+        if len(case_identifiers) <= num_images_properties_loading_threshold:
+            for i in self.dataset.keys():
+                self.dataset[i]['properties'] = load_pickle(self.dataset[i]['properties_file'])        
```

Comment:
- nUNet has a functionality to train cascade models, where you have a first stage model which is low resolution and a second stage model that takes in input CT and the low resolution segmentation mask and predicts high resolution segmentation mask. The "seg_from_prev_stage_file" is to facilitate that. We don't need it as of now and it can be removed. 
- num_images_properties_loading_threshold - This is a variable they use to decide whether to load the input image properties into RAM or read them on the fly from the pickle file. The properties contain info such as where the foreground pixels are, the orginal size of the image and other meta data. The default value of this is kept at 0, which means nnUNet always reads metadata (most useful of which is foreground pixel location) on the fly. I don't see the value in having this either and am ok making a choice about reading metadata on the fly or loading into RAM and keeping that fixed. 
- I don't understand the last q

Reply:
Moved this action to a separate function, but not deleted the sections for (1) and (2). I have added comments explaining what it does though.

Reply:
Last question was answered by your response that explains `num_images_properties_loading_threshold`